validations/webhook.test.ts

```diff
@@ -0,0 +1,336 @@
+import { AddKycParams, FiatConnectClient } from '@fiatconnect/fiatconnect-sdk'
+import {
+  CryptoType,
+  FiatType,
+  KycSchema,
+  KycStatus,
+  Network,
+  TransferStatus,
+  TransferType,
+} from '@fiatconnect/fiatconnect-types'
+import { ethers } from 'ethers'
+import { config } from '../src/config'
+import { MOCK_FIAT_ACCOUNTS } from '../src/mock-data/fiat-account'
+import { MOCK_KYC } from '../src/mock-data/kyc'
+import { MOCK_QUOTE } from '../src/mock-data/quote'
+import { expect, use } from 'chai'
+import { chaiPlugin } from 'api-contract-validator'
+import chaiAsPromised from 'chai-as-promised'
+import { randomUUID } from 'crypto'
+import path from 'path'
+import { checkObjectAgainstModel } from '../src/check-response-schema'
+import axios from 'axios'
+import {
+  WEBHOOK_RECIPIENT_BASE_URL,
+  WEBHOOK_RECIPIENT_HISTORY_ENDPOINT,
+} from '../src/constants'
+
+const apiDefinitionsPath = path.join(config.openapiSpec)
+use(chaiPlugin({ apiDefinitionsPath }))
+use(chaiAsPromised)
+
+interface WebhookTransferEventRecord {
+  provider: string // the field in the body sent by the provider
+  event_id: string
+  address: string
+  status: TransferStatus
+  transfer_type: TransferType
+  fiat_type: FiatType
+  crypto_type: CryptoType
+  amount_provided: string
+  amount_received: string
+  fee: string
+  fiat_account_id: string
+  transfer_id: string
+  transfer_address: string
+  event_timestamp: string
+  timestamp: string
+  provider_id: string // the URL path param, valora defined
+}
+
+interface WebhookKycStatusEventRecord {
+  event_id: string
+  provider: string // the field in the body sent by the provider
+  address: string
+  kyc_schema: string
+  kyc_status: string
+  event_timestamp: string
+  timestamp: string
+  provider_id: string // the URL path param, valora defined
+}
+
+async function validateTransferWebhook(
+  params: { transferId: string; address: string; status: TransferStatus },
+  retries: number,
+): Promise<void> {
+  const { transferId, address, status } = params
+  if (retries < 0) {
+    throw new Error('Webhook event could not be found')
+  }
+  await new Promise((resolve) => setTimeout(resolve, 2000))
```

Comment:
seems like this stuff could be a bit more DRY if shared between the validation helpers. also, there's a nice [NPM library for retries with backoff](https://www.npmjs.com/package/exponential-backoff). `timeMultiple` can be set to 1 for a constant retry time interval.

Reply:
not that important though.